### PR TITLE
Working stm32f4 nucleo + wifi + sensor board demo

### PR DESCRIPTION
**Pull request by Palantir555**
There are a couple details to implement, such as disabling and deinitializing sensors, detection for accelerometer interrupts (taps, etc.) and some error handling in main.c.

Most changes in main.c are just indentation/format normalisation by `clang-format`. Sorry it made a mess of the `diff` 😅  I'll use a separate branch next time. The only relevant modifications to `main.c` were:

1. Addition of the Magnetometer
2. Moved the EXTI interrupt handler into main.c
3. Wrote a new pub_button_interrupt function for better responsiveness in the product simulator
4. Wrote functional publishers and on_led callback

### CODE REVIEW

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c [hunk 1]
```diff
@@ -342,7 +362,7 @@ int8_t io_read_humidity( float* read_value )
  * @param  dec_prec the decimal precision to be used
  * @retval None
  */
-void floatToInt( float in, int32_t* out_int, int32_t* out_dec, int32_t dec_prec )
+static void floatToInt( float in, int32_t* out_int, int32_t* out_dec, int32_t dec_prec )
```
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c]: :+1: 
Just one small note here - I wouldn't call it a `floatToInt` cause what it does it just splits the float number into two pieces. So maybe a `floatSplit` or something like that ? `FloatToInt` suggests a cast so it's misleading - at least for a game programmer :)

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c [hunk 2]
```diff
@@ -355,3 +375,60 @@ void floatToInt( float in, int32_t* out_int, int32_t* out_dec, int32_t dec_prec
...
+ * @retval  0 = Success
+ * @retval -1 = Error
+ */
+int8_t io_float_to_string( float input, char* buf , int32_t buf_size )
+{
+    int retv = 0;
+    int32_t input_integer = 0, input_fractional = 0;
+    floatToInt( input, &input_integer, &input_fractional, 2 );
+    retv = snprintf( buf, buf_size, "%ld.%02ld", input_integer, input_fractional );
+    if ( retv > buf_size )
```
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c]: :+1:

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c [hunk 3]
```diff
@@ -355,3 +375,60 @@ void floatToInt( float in, int32_t* out_int, int32_t* out_dec, int32_t dec_prec
...
+ * @param  Bytes allocated for *buf
+ * @retval  0 = Success
+ * @retval -1 = Error
+ */
+int8_t io_axes_to_json( SensorAxes_t axes, char* buf , int32_t buf_size )
+{
+    int retv = 0;
+    retv = snprintf( buf, buf_size, "{\"x\": %ld, \"y\": %ld, \"z\": %ld}",
+                     axes.AXIS_X, axes.AXIS_Y, axes.AXIS_Z );
+    if ( retv > buf_size )
```
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c]: :+1:

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 4]
```diff
@@ -130,469 +138,565 @@ static xi_state_t init_xively_topics( xi_context_handle_t in_context_handle );
...
-xi_state_t pub_gyroscope( const mqtt_topic_descr_t* const mqtt_topic_descr )
+xi_state_t pub_barometer( const mqtt_topic_descr_t* const mqtt_topic_descr )
 {
-	return xi_publish( gXivelyContextHandle,
-			mqtt_topic_descr->name, "greetings from gyroscope",
-			XI_MQTT_QOS_AT_MOST_ONCE,
-			XI_MQTT_RETAIN_FALSE, NULL, NULL );
+    char out_msg[IO_FLOAT_BUFFER_MAX_SIZE] = "";
+    float sensor_input                     = 0;
+    io_read_pressure( &sensor_input );
```
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: Just a question - Is there a reason not to check for the return value state of the read and convertion functions ?

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 5]
```diff
@@ -130,469 +138,565 @@ static xi_state_t init_xively_topics( xi_context_handle_t in_context_handle );
...
+    char* payload;
+    ( io_read_button() ) ? ( payload = "1" ) : ( payload = "0" );
+    return xi_publish( gXivelyContextHandle, mqtt_topic_descr->name, payload,
+                       XI_MQTT_QOS_AT_MOST_ONCE, XI_MQTT_RETAIN_FALSE, NULL, NULL );
+}
+
+void pub_button_interrupt( void )
+{
+    /* TODO: Verify we are connected to Xively before calling xi_publish */
+    char payload[] = "1";
```
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: `const`

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c (hunk 2 above)
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c]: I think the condition should be ```retv >= buf_size``` according to the documentation.

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c (hunk 3 above)
Comment (olgierdh) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c]: I think the condition should be ```retv >= buf_size``` according to the documentation.

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c (hunk 1 above)
Comment (Palantir555) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/demo_io.c]: Good point. I lifted that function from somewhere in the MEMS SDK, name and all. I'll rename it :)

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c (hunk 4 above)
Comment (Palantir555) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: I'll get to that soon; this was just the first version of a working demo. I still need to add some error handling, clean up main.py and fix some edge cases.

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c (hunk 5 above)
Comment (Palantir555) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: :+1:

Regarding that TODO; is there a recommended way to query libxively for its connection status? Or do I need to keep track of it at the application level?

**Comment by Palantir555** [comment]
@olgierdh Thanks for the review! All changes implemented.  I think this branch is ready for merger. I'll use a new branch for the additional error handling, clean-up, etc.

**Pull request merged**